server/src/main/java/org/elasticsearch/action/admin/cluster/coordination/ClusterFormationInfoAction.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.admin.cluster.coordination;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.ActionRequest;
+import org.elasticsearch.action.ActionRequestValidationException;
+import org.elasticsearch.action.ActionResponse;
+import org.elasticsearch.action.ActionType;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.HandledTransportAction;
+import org.elasticsearch.cluster.coordination.ClusterFormationFailureHelper;
+import org.elasticsearch.cluster.coordination.Coordinator;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.transport.TransportService;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * This class is used to fetch the ClusterFormationState from another node. The ClusterFormationState provides information about why that
+ * node thinks that cluster formation has failed.
+ */
+public class ClusterFormationInfoAction extends ActionType<ClusterFormationInfoAction.Response> {
+
+    public static final ClusterFormationInfoAction INSTANCE = new ClusterFormationInfoAction();
+    public static final String NAME = "cluster:internal/formation/info";
+
+    private ClusterFormationInfoAction() {
+        super(NAME, ClusterFormationInfoAction.Response::new);
+    }
+
+    public static class Request extends ActionRequest {
+
+        public Request() {}
```

Comment:
We'll have to request this information from all master eligible nodes so the request should be able to express that?

Reply:
Okido, sounds good